astropy/config/paths.py

```diff
@@ -165,14 +166,33 @@ def wrapper(*args: P.args, **kwargs: P.kwargs) -> None:
 
     @classmethod
     def _get_dir_path(cls, rootname: str) -> Path:
+        if (usrpth_str := os.getenv(cls._astropy_env_dir)) is not None:
+            usrpth = Path(usrpth_str).expanduser()
+
+            if usrpth.is_file():
+                raise FileExistsError(
+                    f"Cannot create directory under {usrpth_str} as requested "
+                    f"via the environment variable {cls._astropy_env_dir}: "
+                    "a file with this name already exists."
+                )
+
+            if not usrpth.parent.exists():
+                raise FileNotFoundError(
+                    f"Cannot create directory under {usrpth_str} as requested "
+                    f"via the environment variable {cls._astropy_env_dir}: "
+                    f"the parent directory {usrpth.parent} does not exist."
+                )
+            usrpth.mkdir(exist_ok=True)
+            return usrpth.resolve()
+
         if (xch := cls._temp_path) is not None:
```

Comment:
`cls._temp_path` being handled after the Astropy environment variables means that `set_temp_cache` and `set_temp_config` will fail silently at runtime if the corresponding Astropy environment variable is used. The fact that this is not being caught be the tests means that we need to add tests that would check which directory customization mechanism has priority if multiple are used simultaneously.

Reply:
good point